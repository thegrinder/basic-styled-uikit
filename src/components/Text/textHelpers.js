import { css } from 'styled-components';

export const getTextSizing = (theme, sizing) => (
  theme.uiKit.text.sizings[sizing]
);
export const textSizingStyle = ({ theme, sizing }) => (
  css`font-size: ${getTextSizing(theme, sizing)};`
);

export const getTextColor = (theme, colour) => (
  theme.uiKit.text.colours[colour]
);
export const textColorStyle = ({ theme, colour }) => (
  css`color: ${getTextColor(theme, colour)};`
);
