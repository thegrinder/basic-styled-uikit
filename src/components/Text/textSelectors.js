import { getText } from '../../theme/themeSelectors';

export const getTextSizingStyle = (theme, sizing) => getText(theme).sizings[sizing];

export const getTextColor = (theme, color) => getText(theme).colors[color];

export const getTextCommonStyle = theme => getText(theme).common;
