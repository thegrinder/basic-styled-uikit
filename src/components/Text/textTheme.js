import { rem } from '../../helpers/utils';

const textTheme = {
  colours: {
    white: '#fff',
    lightest: '#ccc',
    light: '#999',
    normal: '#666',
    dark: '#424242',
    primary: '#1e87f0',
    success: '#32d296',
    warning: '#faa05a',
    danger: '#f0506e',
  },
  sizings: {
    small: rem(14),
    medium: rem(16),
    large: rem(20),
  },
};

export default textTheme;
