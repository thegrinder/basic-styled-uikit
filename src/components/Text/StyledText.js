import styled, { css } from 'styled-components';
import { string, bool } from 'prop-types';

import { getBaseLineHeight, calcLineHeight } from '../../theme/typography';
import { getTextSizingStyle, getTextColor, getTextCommonStyle } from './textSelectors';

const propTypes = {
  color: string.isRequired,
  sizing: string.isRequired,
  marginBottom: bool.isRequired,
};

const StyledText = styled.span`
  display: block;
  ${({
    theme,
    color,
    sizing,
    marginBottom,
  }) => {
    const baseLineHeight = getBaseLineHeight(theme);
    const sizingStyle = getTextSizingStyle(theme, sizing);
    return css`
      ${getTextCommonStyle(theme)}
      color: ${getTextColor(theme, color)};
      line-height: ${calcLineHeight(sizingStyle.fontSize, baseLineHeight)};
      margin-bottom: ${marginBottom ? `${getBaseLineHeight(theme)}rem` : '0'};
      ${sizingStyle}
    `;
  }}
`;

StyledText.propTypes = propTypes;

export default StyledText;
