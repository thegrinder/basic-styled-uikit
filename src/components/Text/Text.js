import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledText from './StyledText';

const propTypes = {
  /** determines the size of the text; one of: l, m, s, xs */
  sizing: string,
  /**
   * one of: neutral0, neutral50, neutral100, neutral200, neutral300, neutral400,
   * neutral500, neutral600, neutral700, neutral800, neutral900, neutral1000,
   * primary500, success500, danger500, warning500
   * */
  color: string,
  /** rendered html tag */
  as: oneOf(['span', 'p']),
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const defaultProps = {
  sizing: 'm',
  color: 'default',
  as: 'span',
  marginBottom: false,
};

const Text = ({
  sizing,
  color,
  marginBottom,
  as,
  ...rest
}) => (
    <StyledText
      as={as}
      marginBottom={marginBottom}
      sizing={sizing}
      color={color}
      {...rest}
    />
);

Text.propTypes = propTypes;
Text.defaultProps = defaultProps;

export default Text;
