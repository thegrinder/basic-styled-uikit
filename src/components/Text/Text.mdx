---
name: Text
menu: Components
---

import { Playground, Props } from 'docz';
import Text from './Text';
import ColorBox from '../ColorBox/ColorBox';


# Text

<Props of={Text} />


### Text tags

<Playground>
  <Text as="span">Span tag</Text>
  <Text as="p">Paragraph tag</Text>
</Playground>


### Text sizes

<Playground>
  <Text sizing="l">Large</Text>
  <Text sizing="m">Medium</Text>
  <Text sizing="s">Small</Text>
  <Text sizing="xs">Extra small</Text>
</Playground>


### Text colors

<Playground>
  <Text color="primary">primary</Text>
  <Text color="danger">danger</Text>
  <Text color="success">success</Text>
  <Text color="warning">warning</Text>
  <Text color="neutral" emphasis="high">neutral emphasis high</Text>
  <Text color="neutral" emphasis="normal">neutral emphasis normal</Text>
  <Text color="neutral" emphasis="low">neutral emphasis low</Text>

</Playground>


### Text with a bottom margin

<Playground>
  <Text marginBottom={true}>Text with margin</Text>
  <Text marginBottom={true}>Text with margin</Text>
</Playground>