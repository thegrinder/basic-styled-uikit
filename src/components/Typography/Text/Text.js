import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledText from './StyledText';
import { useColorBoxContext } from '../../ColorBox/context';

const propTypes = {
  /** determines the size of the text; one of: l, m, s, xs */
  sizing: string,
  /** one of: neutral, primary, success, danger, warning */
  color: string,
  /** one of: high, normal, low */
  emphasis: string,
  /** determines whether colors defined for the background are ignored */
  ignoreBackground: bool,
  /** rendered html tag */
  as: oneOf(['span', 'p']),
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const defaultProps = {
  sizing: 'm',
  color: 'neutral',
  ignoreBackground: false,
  as: 'span',
  marginBottom: false,
  emphasis: 'high',
};

const Text = ({
  sizing,
  color,
  ignoreBackground,
  marginBottom,
  emphasis,
  as,
  ...rest
}) => (
  <StyledText
    as={as}
    marginBottom={marginBottom}
    sizing={sizing}
    color={color}
    emphasis={emphasis}
    ignoreBackground={ignoreBackground}
    bgColor={useColorBoxContext()}
    {...rest}
  />
);

Text.propTypes = propTypes;
Text.defaultProps = defaultProps;

export default Text;
