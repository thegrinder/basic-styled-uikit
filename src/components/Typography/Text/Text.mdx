---
name: Text
menu: Components
---

import { Playground, Props } from 'docz';
import Text from './Text';
import ColorBox from '../../ColorBox/ColorBox';


# Text

<Props of={Text} />


### Text tags

<Playground>
  <Text as="span">Span tag</Text>
  <Text as="p">Paragraph tag</Text>
</Playground>


### Text sizes

<Playground>
  <Text sizing="l">Large</Text>
  <Text sizing="m">Medium</Text>
  <Text sizing="s">Small</Text>
  <Text sizing="xs">Extra small</Text>
</Playground>


### Text colors

<Playground>
  <Text color="primary">primary</Text>
  <Text color="danger">danger</Text>
  <Text color="success">success</Text>
  <Text color="warning">warning</Text>
  <Text color="neutral" emphasis="high">neutral emphasis high</Text>
  <Text color="neutral" emphasis="medium">neutral emphasis medium</Text>
  <Text color="neutral" emphasis="low">neutral emphasis low</Text>

</Playground>

### Text colors on different backgrounds

<Playground>
  <ColorBox bgColor="primary">
    <Text>neutral on primary background</Text>
  </ColorBox>
  <ColorBox bgColor="success">
    <Text>neutral on success background</Text>
  </ColorBox>
  <ColorBox bgColor="danger">
    <Text>neutral on danger background</Text>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Text>neutral on warning background</Text>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Text ignoreBackground={true} color="primary">
      background ignored - primary text
    </Text>
  </ColorBox>
</Playground>


### Text with a bottom margin

<Playground>
  <Text marginBottom={true}>Text with margin</Text>
  <Text marginBottom={true}>Text with margin</Text>
</Playground>