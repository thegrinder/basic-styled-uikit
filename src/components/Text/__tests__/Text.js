import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';

import theme from '../../../theme/theme';
import Text from '../Text';


const children = 'children';

const renderComponent = props => render(
  <ThemeProvider theme={theme}>
    <Text {...props}>{children}</Text>
  </ThemeProvider>,
);

describe('<Text />', () => {
  afterEach(cleanup);

  it('should render correctly with default props', () => {
    const { getByText } = renderComponent();
    const component = getByText(children);
    expect(component).toBeDefined();
    expect(component).toHaveTextContent(children);
    expect(component).toMatchSnapshot();
  });

  it('should render correctly with custom props', () => {
    const { getByText } = renderComponent({
      sizing: 'xs',
      color: 'primary',
      marginBottom: true,
    });
    const component = getByText(children);
    expect(component).toBeDefined();
    expect(component).toHaveTextContent(children);
    expect(component).toMatchSnapshot();
  });

  it('should render <span> tag by default', () => {
    const { getByText } = renderComponent();
    const component = getByText(children);
    expect(component.tagName).toEqual('SPAN');
    expect(component).toMatchSnapshot();
  });

  it('should render render <p> tag', () => {
    const { getByText } = renderComponent({ as: 'p' });
    const component = getByText(children);
    expect(component.tagName).toEqual('P');
    expect(component).toMatchSnapshot();
  });
});
