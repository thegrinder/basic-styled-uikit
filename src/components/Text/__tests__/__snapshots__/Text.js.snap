// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should render <span> tag by default 1`] = `
<span
  class="StyledText-r7l7li-0 bHYFxY"
  color="default"
>
  children
</span>
`;

exports[`<Text /> should render correctly with custom props 1`] = `
<span
  class="StyledText-r7l7li-0 cCETgl"
  color="primary"
>
  children
</span>
`;

exports[`<Text /> should render correctly with default props 1`] = `
<span
  class="StyledText-r7l7li-0 bHYFxY"
  color="default"
>
  children
</span>
`;

exports[`<Text /> should render render <p> tag 1`] = `
<p
  class="StyledText-r7l7li-0 bHYFxY"
  color="default"
>
  children
</p>
`;
